### ISSUE TITLE
New user: Support for adaptation on mixed element types meshes

### ISSUE THREAD

**Issue by a-jp** [mfem/mfem#1071]
Hi,
Hope it's ok to use the issue tracker to post questions?  I'm new to MFEM (only compiled it at the weekend).  

In 2D for an unstructured mixed cell type mesh (triangles/quads) is there support for adaption of each type (triangles or quads) independently within the same mesh? Does this allow for hanging nodes being generated in mixed element type meshes?  Likewise same question in 3D for tets/hexs?  Or, do meshes need to have a homogenous cell type for adaption?

For unstructured mixed element cell type meshes, with the exception of hypre/metis/mpi what additional packages should I enable to allow for support of unstructured mixed cell type meshes that I then wish to subsequently adapt (assuming the above is confirmed possible)?

Thanks,
Andy

**Comment by jakubcerveny** [comment]
Hi @a-jp, the features you are requesting are fairly new and have not been merged into master yet:
 - `nc-prism-dev` (#713) adds support for mixed nonconforming meshes, including quads/triangles and hexes/prisms
 - `nc-tet-dev` (#949) further addes support for nonconforming octasection of tetrahedra and mixed hex/prism/tet meshes

@v-dobrev, it would be great to get your review of #713...

**Comment by a-jp** [comment]
Many thanks for the answer.  So it would be correct to say that if I tried to refine one quad in a mixed quad/triangle mesh I would expect an error at present?  Your point is that the above gives a route to that soon?

> For unstructured mixed element cell type meshes, with the exception of hypre/metis/mpi what additional packages should I enable to allow for support of unstructured mixed cell type meshes that I then wish to subsequently adapt (assuming the above is confirmed possible)?

Do you have any comments to the above?

Thanks,
Andy

**Comment by jakubcerveny** [comment]
@a-jp Did you have any luck with the `nc-prism-dev` branch? Can we close this issue?

**Comment by a-jp** [comment]
I got slightly side tracked...  I have been working with the help of @cwsmith to develop a reader and a writer for apf/pumi to read and write in parallel CGNS meshes.  I'm close to pushing this and setting up a PR to request a merge.  I was waiting for that to be done so that I could use the pumi-apf bridge in MFEM to get my meshes in. After I did that I was going to try adapting them.  Outside of pumi/apf I have written a serial CGNS reader for MFEM with no requirement to build pumi/apf interface which appears to work.  Is that of interest to push, or would you see it better to just access that reader implicitly via the pumi-apf bridge that exists and the impending PR?  

I have a strong requirement to be able to adapt and refine mixed cell meshes containing (tri/quads in 2D and tets/hexs in 3D) but I reckon by the time I get the above finished this will all be merged anyway.  I need #713 and #949 for my use case to work.

Short answer please close.

Andy

**Comment by jakubcerveny** [comment]
The first part is probably for @tzanio to answer. As for #713 and #949 we're hoping to get those merged soon too.

**Comment by tzanio** [comment]
> Outside of pumi/apf I have written a serial CGNS reader for MFEM with no requirement to build pumi/apf interface which appears to work. Is that of interest to push, or would you see it better to just access that reader implicitly via the pumi-apf bridge that exists and the impending PR?

That can certainly be of interest. Can you open a PR so we can take a look?
